[Doc] Refactor global settings (#6071)
taichi-dev/taichi: docs/lang/articles/reference/global_settings.md

# Global Settings

> The call `ti.init()` is indispensable in every Taichi program. It allows you to customize your Taichi runtime by passing arguments to it or setting environment variables. Each argument or environment variable controls one specific behavior of the Taichi runtime. For example, the argument `arch` specifies the backend, and the argument `debug` decides whether to run the program in debug mode. The document [Hello, World!](../get-started/index.md) gives a brief introduction to this function, and this document provides more details.

Reviewer: ```suggestion
The call `ti.init()` is indispensable in every Taichi program. Taichi allows you to customize your Taichi runtime by passing arguments to `ti.init()` or setting environment variables. Each argument or environment variable controls one specific behavior of the Taichi runtime. For example, the argument `arch` specifies the backend, and the argument `debug` decides whether to run the program in debug mode. The document [Hello, World!](../get-started/index.md) gives a brief introduction to this function, and this document provides more details.
```

> Generally, when calling `ti.init()`, Taichi executes the following steps to initialize a specific configuration. We use the `arch` argument as an example:

Reviewer: ```suggestion
Generally, when you call `ti.init()`, Taichi executes the following steps to initialize a specific configuration. We use the `arch` argument as an example:
```
'calling' takes 'Taichi' as its subject, which is incorrect.

> 1. Firstly, Taichi looks for the argument that is passed to `ti.init()`. In this case, after Taichi reads `ti.init(arch=cuda)`, it chooses CUDA as the backend and omits the environment variable.

Reviewer: ```suggestion
1. Taichi looks for the arguments passed to `ti.init()`. In this case, after Taichi reads `ti.init(arch=cuda)`, it chooses CUDA as the backend and omits the environment variable.
```
I believe argument should be in plural form here.

> 2. If no argument is found, Taichi checks the corresponding environment variable. In this case, if `arch` is not specified but replaced by the environment variable `export TI_ARCH=cuda`, Taichi still chooses CUDA as the backend.

Reviewer: ```suggestion
2. If no argument is found, Taichi checks the corresponding environment variable. In this case, if `arch` is not specified but the environment variable is set to `export TI_ARCH=cuda`, Taichi chooses CUDA as the backend.
```

> 3. If no customized setting is found, Taichi uses a default configuration. In this case, if neither the argument `arch` is specified nor the environment variable `TI_ARCH` is found, Taichi adopts the default backend `arch=ti.cpu`.

Reviewer: ```suggestion
3. If no customized setting is found, Taichi uses a default configuration. In this case, if neither the argument `arch` is specified nor an environment variable `TI_ARCH` is found, Taichi adopts the default backend `arch=ti.cpu`.
```

> Below are some most frequently used configurations you can set with the `ti.init()` call:

Reviewer: ```suggestion
Following are some frequently used configurations that the `ti.init()` method call supports:
```

# Global Settings

The call `ti.init()` is indispensable in every Taichi program. It allows you to customize your Taichi runtime by passing arguments to it or setting the corresponding environment variables. Each argument or environment variable will control one specific behavior of the Taichi runtime. For example, the argument `arch` for specifying the backend, the argument `debug` for running the program in debug mode. We have introduced this function in our very first article ["getting started"](../get-started/index.md), but what given there was only a coarse sketch. In this article we will show its full functionalities.

> In summary, when calling `ti.init()`, Taichi will execute the following steps to initialize a specific configuration: (we use the `arch` argument to illustrate this)

Reviewer 2: ```suggestion
Generally, when calling `ti.init()`, Taichi executes the following steps to initialize a specific configuration. We use the `arch` argument as an example:
```

[Compilation Options]

    advanced_optimization: bool
        Enable/disable advanced optimization to save compile time & possible errors.

    fast_math: bool
        Enable/disable fast math to prevent possible undefined math behavior.

    print_ir: bool
        Turn on/off printing intermediate IR generated.

[Runtime Options]

    cpu_max_num_threads: int
        Set the number of threads used by the CPU thread pool.

    debug: bool
        Run program in debug mode.

    default_cpu_block_dim: int
        Set the number of threads in a block on CPU.

    default_gpu_block_dim: int
        Set the number of threads in a block on GPU.

    default_fp: [ti.f32, ti.f64]
        Set the default precision for floating-point numbers in the Taichi scope.

    default_ip: [ti.i32, ti.i64]
        Set the default precision for integers in the Taichi scope.

    dynamic_index: bool
        Enable/disable using variables as index to access vector/matrix elements in the Taichi scope.

    kernel_profiler: bool
        Turn on/off kernel performance profiling.

    offline_cache: bool
        Enable/disable offline cache of the compiled kernels.

    offline_cache_file_path: str

>         Directory holding the offline cached files.

Author: @Olinaaaloompa Should change set a directory -> specify a directory

    packed: bool
        Enable/disable the packed memory layout.

    random_seed: int
        Set a custom seed for the random number generator.

[Logging Options]

    log_level: [ti.INFO, ti.TRACE, ti.WARN, ti.ERROR, ti.CRITICAL, ti.DEBUG]
        Set the logging level.

    verbose: bool
        Eliminate verbose outputs. e.g. `ti.init(verbose=False)`.

[Develop Options]

    gdb_trigger: bool

>         Enable/disable triggering GDB when Taichi crashes, e.g. `ti.init(gdb_trigger=True)`.

Author: I'm not sure the word "silence" is a good choice here.

```

**Customize via environment variables**

> Below are some environment variables that you can set to customize your Taichi program, they overlap but are not in one-to-one correspondence with the `ti.init()` arguments listed above:

Author: @Olinaaaloompa  The meaning of the original version is changed: some environment settings do not map to `init` arguments.

> Below are some environment variables that you can set to customize your Taichi program, they overlap but are not in one-to-one correspondence with the `ti.init()` arguments listed above:

Reviewer 2: Have rephrased. Please check the sentence again.

> Below are some environment variables that you can set to customize your Taichi program, they overlap but are not in one-to-one correspondence with the `ti.init()` arguments listed above:

Author: "Interchangeable" is a bit obscure here. Let's say "not all of them have corresponding arguments of the `init` function"?
